types/react-navigation/index.d.ts

```diff
@@ -86,7 +87,7 @@ export type NavigationLeafRoute<Params> = {
    * Params passed to this route when navigating to it,
    * e.g. `{ car_id: 123 }` in a route that displays a car.
    */
-  params?: Params,
+  params: Params,
```

Comment:
Wow, so sorry, you're totally right. I was mistakenly thinking of this as still being `params?: NavigationParams`, not the typed-generic `Params` (which is even more ironic since I made this change not too long ago in https://github.com/phanalpha/DefinitelyTyped/commit/9be2563bfc81f9127474f0c319c54532a3321517). This should indeed be changed according to what you did, since this is a typed generic, and the user should in general be responsible for adding `undefined if they want. But if this were still the generic *map* `NavigationParams`, then we would have wanted to leave this as `params?:`.

Really sorry about that, I've done some really sloppy and rushed work reviewing your PR. But thanks for the patience and for keeping up the explanation of your approaches and changes!

The only thing here is that unfortunately I believe my example with the empty `state.params` could still be applicable if end-users don't take care. I've copied it below:

```
interface ThisComponentsNavigationParamsInterface {
    something?: string,
}

...

render() {
    if (this.props.navigation.state.params.something) {
       return ( <Text>...</Text> );
    } else {
       return ( <View>...</View> );
    }
}

...

<From some other component>
this.props.navigation.navigate(ThisComponentRouteName);
```

Technically, what should probably be done instead is:
```
const params: ThisComponentsNavigationParamsInterface = {};
this.props.navigation.navigate(ThisComponentRouteName, params);
```
But this may be a gotcha for new users who are more loosely using the typed, and who would still expect at least an empty object in `this.props.navigation.state.params`. Thoughts there?